### ISSUE TITLE
Add ZocDoc

### ISSUE THREAD

**Issue by harcod** [livgust/covid-vaccine-scrapers#171]
This is the top-level page: https://www.zocdoc.com/vaccine

This appears to be Massachusetts:
https://www.zocdoc.com/vaccine/search/MA?flavor=state-search

Here is the original link supplied by @livgust (although this goes to a 404 right now)
https://www.zocdoc.com/wl/tuftscovid19vaccination/practice/64825?reason_visit=5243

This is the top-level section for Tufts, but the above link
https://www.tuftsmcvaccine.org/

Tufts is apparently separate from the general ZocDoc  link above.

**Comment by harcod** [comment]
![image (1)](https://user-images.githubusercontent.com/546007/112061419-dc914400-8b34-11eb-8130-e464fc67d492.png)

I have the associated HTML file and can send that to you. Just ask.

**Comment by pfurbacher** [comment]
Starting at [Tufts Vaccine Scheduling front page](https://www.tuftsmcvaccine.org/), I started digging into the XHR network chatter which results from navigating to the calendar page. What I think are the important pieces are written up in the following:

[zodoc-tufts notes.txt](https://github.com/livgust/covid-vaccine-scrapers/files/6209013/zodoc-tufts.notes.txt)

If those of you who are cleverer at this than I am can read through the notes and perhaps make suggestions as to where to go from here, I'd appreciate it. Or if someone wants to take it from here, that's good too.

I had been looking at the other Zocdoc sites (AFC Urgent Care and Holtzman Medical Group - Mount Ida Campus) and actually tried assembling such a node-fetch request from the bits and pieces (cookies, body, url, etc.), before discovering Chrome's "copy as Node.js fetch". My attempt looked pretty the same as what you see in the above notes. I wasn't successful, so there's something I'm not understanding about this way of teasing data responses.

**Comment by livgust** [comment]
@pfurbacher I found the same thing as you - that gql POST should be all we need for Tufts. I'm a little confused what you're looking for and what you're saying about the other locations. Are you having trouble getting the data for all of them, or just the ones that aren't Tufts?

I tried the fetch query you had in your notes on plain ol' google.com (from the console in the dev tools) and it failed, but when I ran it from the console in the zocdoc web page it worked, so there's some CORS stuff going on. I know there are ways to fix it but I'm really bad at understanding that stuff haha...

**Comment by pfurbacher** [comment]
> Are you having trouble getting the data for all of them, or just the ones that aren't Tufts?

I was looking at the other Zocdoc sites (AFC UC and Holtzman) and didn't have success getting data. (Can't remember what the error was -- I'll dig up the code and report back later.) With this Tufts issue, I was just going through the newest issues and noticed Rob had made some progress with a Zocdoc-powered site, so I decided to look at this as well. Since I had failed with the earlier fetch attempt, I thought I'd ask for ideas from others for pointers. That is, there are other scrapers that first get cookies/authorization, then make other requests to eventually get data -- I was hoping that those who contributed such scrapers might have ideas, or know better how to do this. (Gosh, I hope what I'm writing makes sense.)

I don't know anything about CORS, so I can't help with that.

**Comment by relm923** [comment]
```sh
# Search for locations by state (MA)
curl 'https://api.zocdoc.com/directory/v2/gql' \
  -H 'content-type: application/json' \
  --data-raw $'{"operationName":"search","variables":{"includeCustomProviderProps":false,"includeSpo":false,"insuranceCarrierId":"-1","insurancePlanId":"-1","parameters":{"directoryId":"-1","flavor":"state-search","insuranceCarrierId":"-1","insurancePlanId":"-1","itemOffset":0,"itemQuantity":17,"latencyTestDelayMs":0,"url":"https://www.zocdoc.com/vaccine/search/MA?flavor=state-search","pageType":"SearchPage","pageId":"45f795f2b9815f19689fbcada1553e82","procedureId":"5243","rankBy":"Default","searchLocation":{"address":"MA","coordinates":null,"ip":"66.31.16.189","placemarkData":{"state":"MA"}},"searchQueryGuid":"ce08c5cf-bb97-44ea-9ffe-552435cf36b7","sessionId":"0dc4eb2a1d0b47069bb183b233f88bc2","specialtyId":"510","timeFilter":"AnyTime","trackingId":"798d9a10-b1c5-40e4-b5e8-445c31eb77a9","locale":"en","virtualLocationPageSize":7,"visitType":"inPersonVisit","requestedFacets":["procedures","specialties","sex","hospital_affiliations","languages","sees_children","time_of_day","day_availability","distance_radius"],"excludePreviewProviders":true,"filters":[{"name":"procedures","values":["5243"]}],"callerType":"vaccineSearch"},"specialtyId":"510","includeBadges":false,"includeStrategicId":false,"includeReviews":false,"includeNextAvailability":true,"includeFacets":false,"skipAvailability":true},"query":"query search($directoryId: String, $includeBadges: Boolean\u0021, $includeCustomProviderProps: Boolean\u0021, $includeSpo: Boolean\u0021, $includeReviews: Boolean\u0021, $includeNextAvailability: Boolean\u0021, $includeFacets: Boolean\u0021, $includeStrategicId: Boolean\u0021, $insuranceCarrierId: String\u0021, $insurancePlanId: String\u0021, $isNewPatient: Boolean, $isReschedule: Boolean, $jumpAhead: Boolean, $numDays: Int, $firstAvailabilityMaxDays: Int, $parameters: SearchParameters\u0021, $procedureId: String, $searchRequestId: String, $skipAvailability: Boolean\u0021, $specialtyId: String\u0021, $startDate: String, $timeFilter: TimeFilter, $widget: Boolean) {\\n  insuranceCarrierImage(carrierId: $insuranceCarrierId)\\n  search(parameters: $parameters) {\\n    id\\n    searchResponse {\\n      nearestState\\n      searchRequestId\\n      providerLocations {\\n        ...providerLocation\\n        provider {\\n          properties @include(if: $includeCustomProviderProps) {\\n            key\\n            value\\n            __typename\\n          }\\n          __typename\\n        }\\n        __typename\\n      }\\n      providerLocationsMapDots {\\n        providerId\\n        locationId\\n        latitude\\n        longitude\\n        hasAvailability\\n        __typename\\n      }\\n      facets @include(if: $includeFacets) {\\n        id\\n        facetId\\n        displayName\\n        isMulti\\n        popularOptions {\\n          optionId\\n          displayName\\n          count\\n          __typename\\n        }\\n        selectionRequired\\n        options {\\n          optionId\\n          displayName\\n          count\\n          __typename\\n        }\\n        __typename\\n      }\\n      governmentInsuranceData {\\n        showGovernmentInsuranceWarning\\n        header\\n        message\\n        links {\\n          name\\n          url\\n          __typename\\n        }\\n        __typename\\n      }\\n      totalCount\\n      searchUrl\\n      insuranceCounts {\\n        inNetwork\\n        outOfNetwork\\n        __typename\\n      }\\n      __typename\\n    }\\n    spo @include(if: $includeSpo) {\\n      spoAds {\\n        spoAdDecisionId\\n        adDecisionToken\\n        adRank\\n        providerLocation {\\n          ...providerLocation\\n          __typename\\n        }\\n        __typename\\n      }\\n      telehealthAds {\\n        adDecisionId\\n        decisionToken\\n        provider {\\n          allowedPatient\\n          cost\\n          coupon {\\n            couponCode\\n            discountedCost\\n            __typename\\n          }\\n          fullName\\n          partnerName\\n          providerImage\\n          readableProviderType\\n          branchDeepLink\\n          sourceId\\n          __typename\\n        }\\n        searchResultId\\n        spoServiceAttributionUrl\\n        spoServiceImpressionUrl\\n        __typename\\n      }\\n      __typename\\n    }\\n    searchOrigin {\\n      latitude\\n      longitude\\n      __typename\\n    }\\n    highlightedProviderLocations {\\n      ...providerLocation\\n      __typename\\n    }\\n    __typename\\n  }\\n}\\n\\nfragment providerLocation on ProviderLocation {\\n  id\\n  nextAvailability @include(if: $includeNextAvailability) {\\n    startTime\\n    __typename\\n  }\\n  provider {\\n    id\\n    monolithId\\n    approvedFullName\\n    averageRating\\n    averageBedsideRating\\n    averageWaitTimeRating\\n    canHaveAppointments\\n    frontEndCirclePictureUrl\\n    defaultProcedureId\\n    firstName\\n    isPreview\\n    providerBadges\\n    lastName\\n    mayChargeAdditionalCopay\\n    nameInSentence\\n    onlySeesChildren\\n    postnominal\\n    prenominal\\n    profileUrl\\n    frontEndSquarePictureUrl\\n    status\\n    relevantSpecialty(searchSpecialtyId: $specialtyId) {\\n      id\\n      name\\n      __typename\\n    }\\n    procedures {\\n      id\\n      __typename\\n    }\\n    reviewCount\\n    representativeReview @include(if: $includeReviews) {\\n      comment\\n      __typename\\n    }\\n    specialties {\\n      id\\n      name\\n      __typename\\n    }\\n    badges @include(if: $includeBadges)\\n    visitTypesByState {\\n      allVisitTypes\\n      inPersonVisitOnly\\n      virtualVisitOnly\\n      __typename\\n    }\\n    locations {\\n      id\\n      state\\n      isVirtual\\n      __typename\\n    }\\n    sellingPoints {\\n      id\\n      title\\n      description\\n      imageUrl(extension: svg)\\n      __typename\\n    }\\n    isTopRebookingProvider\\n    highlyRecommendPercentage\\n    lowWaitTimedPercentage\\n    hasNewPatientAvailability\\n    __typename\\n  }\\n  location {\\n    id\\n    monolithId\\n    address1\\n    address2\\n    city\\n    latitude\\n    longitude\\n    name\\n    phone\\n    state\\n    zipCode\\n    isVirtual\\n    virtualVisitType\\n    __typename\\n  }\\n  practice {\\n    id\\n    isUrgentCare\\n    strategicId @include(if: $includeStrategicId)\\n    logo @include(if: $includeStrategicId)\\n    __typename\\n  }\\n  acceptsInsurance(carrierId: $insuranceCarrierId, planId: $insurancePlanId)\\n  distance(unit: \\"mi\\")\\n  searchResultId\\n  ...availability @skip(if: $skipAvailability)\\n  __typename\\n}\\n\\nfragment availability on ProviderLocation {\\n  id\\n  provider {\\n    id\\n    monolithId\\n    __typename\\n  }\\n  location {\\n    id\\n    monolithId\\n    state\\n    phone\\n    __typename\\n  }\\n  availability(directoryId: $directoryId, insurancePlanId: $insurancePlanId, isNewPatient: $isNewPatient, isReschedule: $isReschedule, jumpAhead: $jumpAhead, firstAvailabilityMaxDays: $firstAvailabilityMaxDays, numDays: $numDays, procedureId: $procedureId, searchRequestId: $searchRequestId, startDate: $startDate, timeFilter: $timeFilter, widget: $widget) {\\n    times {\\n      date\\n      timeslots {\\n        isResource\\n        startTime\\n        __typename\\n      }\\n      __typename\\n    }\\n    firstAvailability {\\n      startTime\\n      __typename\\n    }\\n    showGovernmentInsuranceNotice\\n    timesgridId\\n    today\\n    __typename\\n  }\\n  __typename\\n}\\n"}' \
  --compressed \
  | jq '.data.search.searchResponse.providerLocations[].id'

# Provider Location Ids in State
"pr_VUnpWUtg1k2WFBMK8IhZkx|lo_PhFQcSZjdUKZUHp63gDcmx"
"pr_pEgrY3r5qEuYKsKvc4Kavx|lo_f3k2t812AUa9NTIYJpbuKx"
"pr_TeD-JuoydUKqszEn2ATb8h|lo_jbrpfIgELEWWL2j5d3t6Sh"
"pr_iXjD9x2P-0OrLNoIknFr8R|lo_xmpTUhghfUC2n6cs3ZGHhh"
"pr_BDBebslqJU2vrCAvVMhYeh|lo_zDtK7NZWO0S_rgUqjxD1hB"
"pr_4Vg_3ZeLY0aHJJxsCU-WhB|lo_VA_6br22m02Iu57vrHWtaB"
"pr_CUmBnwtlz0C16bif5EU0IR|lo_X6zHHncQnkqyLp-rvpi1_R"

# Search for availability by Provider Location Id
curl 'https://api.zocdoc.com/directory/v2/gql' \
  -H 'content-type: application/json' \
  --data-raw '{"operationName":"providerLocationsAvailability","variables":{"directoryId":"-1","insurancePlanId":"-1","isNewPatient":true,"jumpAhead":true,"numDays":14,"procedureId":"5243","searchRequestId":"7345a02c-6dac-4108-8e76-1ac9f5c0052b","timeFilter":"AnyTime","firstAvailabilityMaxDays":30,"providerLocationIds":["pr_VUnpWUtg1k2WFBMK8IhZkx|lo_PhFQcSZjdUKZUHp63gDcmx","pr_pEgrY3r5qEuYKsKvc4Kavx|lo_f3k2t812AUa9NTIYJpbuKx","pr_TeD-JuoydUKqszEn2ATb8h|lo_jbrpfIgELEWWL2j5d3t6Sh","pr_iXjD9x2P-0OrLNoIknFr8R|lo_xmpTUhghfUC2n6cs3ZGHhh","pr_BDBebslqJU2vrCAvVMhYeh|lo_zDtK7NZWO0S_rgUqjxD1hB","pr_4Vg_3ZeLY0aHJJxsCU-WhB|lo_VA_6br22m02Iu57vrHWtaB","pr_CUmBnwtlz0C16bif5EU0IR|lo_X6zHHncQnkqyLp-rvpi1_R"]},"query":"query providerLocationsAvailability($directoryId: String, $insurancePlanId: String, $isNewPatient: Boolean, $isReschedule: Boolean, $jumpAhead: Boolean, $firstAvailabilityMaxDays: Int, $numDays: Int, $procedureId: String, $providerLocationIds: [String], $searchRequestId: String, $startDate: String, $timeFilter: TimeFilter, $widget: Boolean) {\n  providerLocations(ids: $providerLocationIds) {\n    id\n    ...availability\n    __typename\n  }\n}\n\nfragment availability on ProviderLocation {\n  id\n  provider {\n    id\n    monolithId\n    __typename\n  }\n  location {\n    id\n    monolithId\n    state\n    phone\n    __typename\n  }\n  availability(directoryId: $directoryId, insurancePlanId: $insurancePlanId, isNewPatient: $isNewPatient, isReschedule: $isReschedule, jumpAhead: $jumpAhead, firstAvailabilityMaxDays: $firstAvailabilityMaxDays, numDays: $numDays, procedureId: $procedureId, searchRequestId: $searchRequestId, startDate: $startDate, timeFilter: $timeFilter, widget: $widget) {\n    times {\n      date\n      timeslots {\n        isResource\n        startTime\n        __typename\n      }\n      __typename\n    }\n    firstAvailability {\n      startTime\n      __typename\n    }\n    showGovernmentInsuranceNotice\n    timesgridId\n    today\n    __typename\n  }\n  __typename\n}\n"}' \
  --compressed \
  | jq '.data.providerLocations'

# Availability by Location
[
  {
    "id": "pr_VUnpWUtg1k2WFBMK8IhZkx|lo_PhFQcSZjdUKZUHp63gDcmx",
    "provider": {
      "id": "pr_VUnpWUtg1k2WFBMK8IhZkx",
      "monolithId": "312601",
      "__typename": "Provider"
    },
    "location": {
      "id": "lo_PhFQcSZjdUKZUHp63gDcmx",
      "monolithId": "141505",
      "state": "MA",
      "phone": "(781) 461-0200",
      "__typename": "Location"
    },
    "availability": {
      "times": [
        {
          "date": "2021-03-26",
          "timeslots": [],
          "__typename": "TimeslotDay"
        },
        {
          "date": "2021-03-27",
          "timeslots": [],
          "__typename": "TimeslotDay"
        },
        ...
}

# Open availability example (pulled from WA):
[
  {
    "id": "pr_jS-2Yxu4B0SDDrJJUSFCBh|lo_3ZmaBcByG0ah4L-Pybu09x",
    "provider": {
      "id": "pr_jS-2Yxu4B0SDDrJJUSFCBh",
      "monolithId": "315492",
      "__typename": "Provider"
    },
    "location": {
      "id": "lo_3ZmaBcByG0ah4L-Pybu09x",
      "monolithId": "143227",
      "state": "WA",
      "phone": "(253) 470-8078",
      "__typename": "Location"
    },
    "availability": {
      "times": [
        {
          "date": "2021-03-30",
          "timeslots": [ # POPULATED 'timeslots' DESIGNATE AVAILABLE APPOINTMENTS
            {
              "isResource": false,
              "startTime": "2021-03-30T15:30:00-07:00",
              "__typename": "Timeslot"
            },
            {
              "isResource": false,
              "startTime": "2021-03-30T16:40:00-07:00",
              "__typename": "Timeslot"
            }
          ],
          "__typename": "TimeslotDay"
        },
        ...
}
```

This appears to be working just fine without authentication or any CORS related fixes

**Comment by pfurbacher** [comment]
@relm923 : Thanks for the info. Quite helpful.

**Comment by pfurbacher** [comment]
One note: AFC Urgent Care Springfield will come up as one of the sites, but they say that all vaccinations (when available) will be done at the West Springfield office. We will give the address of the Springfield office, but I'm thinking we should have `extraData` which says "Appointments for the vaccine are only available at AFC Urgent Care West Springfield at this time." As time goes on, we'll have to monitor this piece of advice to see if it changes.

**Comment by pfurbacher** [comment]
I need a sanity check: are these results in the right structure? (Note that `results` is an array of sites. Is that correct?)

```
{
    "version": 1,
    "timestamp": "2021-03-30T021442Z",
    "results": [
        {
            "Tufts Medical Center": {
                "name": "Tufts Medical Center",
                "street": "279 Tremont St",
                "city": "Boston",
                "zip": "02116",
                "signUpLink": "https://www.zocdoc.com/wl/tuftscovid19vaccination/practice/64825?reason_visit=5243",
                "availability": {},
                "hasAvailability": false
            },
            "AFC Urgent Care Dedham": {
                "name": "AFC Urgent Care Dedham",
                "street": "370 Providence Hwy",
                "city": "Dedham",
                "zip": "02026",
                "signUpLink": "https://afcurgentcarededham.com/covid-19-vaccine-registration/",
                "availability": {},
                "hasAvailability": false
            },
            "AFC Urgent Care Worcester": {
                "name": "AFC Urgent Care Worcester",
                "street": "117A Stafford St",
                "city": "Worcester",
                "zip": "01603",
                "signUpLink": "https://afcurgentcareworcester.com/",
                "availability": {},
                "hasAvailability": false
            },
            "AFC Urgent Care New Bedford": {
                "name": "AFC Urgent Care New Bedford",
                "street": "119 Coggeshall St",
                "city": "New Bedford",
                "zip": "02746",
                "signUpLink": "https://afcurgentcarenewbedford.com/",
                "availability": {},
                "hasAvailability": false
            },
            "AFC Urgent Care Saugus": {
                "name": "AFC Urgent Care Saugus",
                "street": "371 Broadway",
                "city": "Saugus",
                "zip": "01906",
                "signUpLink": "https://afcurgentcaresaugus.com/covid-19-vaccination/",
                "availability": {},
                "hasAvailability": false
            },
            "Holtzman Medical Group - Mount Ida Campus": {
                "name": "Holtzman Medical Group - Mount Ida Campus",
                "street": "777 Dedham St",
                "city": "Newton",
                "zip": "02459",
                "signUpLink": "https://www.holtzmanmedical.org/covid19",
                "availability": {},
                "hasAvailability": false
            },
            "AFC Urgent Care West Springfield": {
                "name": "AFC Urgent Care West Springfield",
                "street": "18 Union St",
                "city": "West Springfield",
                "zip": "01089",
                "signUpLink": "https://afcurgentcarespringfield.com/covid-vaccination-registration/",
                "availability": {},
                "hasAvailability": false
            },
            "AFC Urgent Care Springfield": {
                "name": "AFC Urgent Care Springfield",
                "street": "415 Cooley St, Unit 3",
                "city": "Springfield",
                "zip": "01128",
                "signUpLink": "https://afcurgentcarespringfield.com/covid-vaccination-registration/",
                "extraData": "Appointments for the vaccine are only available at AFC Urgent Care West Springfield at this time. ",
                "availability": {},
                "hasAvailability": false
            },
            "latitude": 42.4072107,
            "longitude": -71.3824374
        }
    ]
}
```

**Comment by iann** [comment]
Hi @pfurbacher That's close. I believe the results are supposed to be in the following format.

```
{
    "version": 1,
    "timestamp": "2021-03-30T021442Z",
    "results": [
        {
            "name": "Tufts Medical Center",
            "street": "279 Tremont St",
            "city": "Boston",
            "zip": "02116",
            "signUpLink": "https://www.zocdoc.com/wl/tuftscovid19vaccination/practice/64825?reason_visit=5243",
            "availability": {},
            "hasAvailability": false
        },
        ...
    ]
}
```

**Comment by harcod** [comment]
Don't forget to add the **timestamp** for each location.
It's just `timestamp: new Date(),`

```
{
    "version": 1,
    "timestamp": "2021-03-30T021442Z",
    "results": [
        {
            "name": "Tufts Medical Center",
            "street": "279 Tremont St",
            "city": "Boston",
            "zip": "02116",
            "signUpLink": "https://www.zocdoc.com/wl/tuftscovid19vaccination/practice/64825?reason_visit=5243",
            "availability": {},
            "hasAvailability": false
            "timestamp": "2021-03-30T16:05:30.560Z"
        },
        ...
    ]
}
```

**Comment by livgust** [comment]
closed by #192
